Automattic/wp-calypso
# Pattern Assembler: Use single iframe for the large preview

Related to https://github.com/Automattic/wp-calypso/issues/76020, https://github.com/Automattic/wp-calypso/pull/72364, https://github.com/Automattic/wp-calypso/pull/76132

## Proposed Changes

As https://github.com/Automattic/wp-calypso/issues/76020 mentioned, there are pros and cons when using a single iframe. This PR is focusing on using a single iframe to see whether the results look good or not.

**The concept of the current approach**

The main idea is to add the dummy overlay onto the hovered pattern on the large preview outside the iframe. So, we can reuse the existing components, styles, etc.

However, it's not easy to detect whether the cursor is leaving the hovered pattern. The reason is we want to keep hovered status when your mouse is over the action bar of the pattern but we cannot detect the actual target when the target is outside of the iframe (we will get `window` instead of the actual one).

[Update on May 4]

Attach the `mouseleave` event on the container of the large preview so we're able to detect whether the cursor is leaving out and we won't need to forward the scroll. Moreover, this way also resolves the performance issue on Firefox!

~So we did a trick to detect the `mouseleave` event on the dummy overlay. But if we did that, we are not able to scroll inside the iframe. Thankfully, Gutenberg has [usePopoverScroll](https://github.com/WordPress/gutenberg/blob/0154f86ed19502bb205a83f0f545058ca824c4c2/packages/block-editor/src/components/block-popover/use-popover-scroll.js#L13) hook to forward the scroll event from their toolbar to the content. Hence, we can simply mimic this approach to forward the scroll event to the iframe.~

With this approach, we can also fix the issue that [the action bar is not sticky to the pattern](pekYwv-1j1-p2) pekYwv-1j1-p2 🎉

Note
* ~It's a bit slow on Firefox~

**Alternatives**

Here are some possible alternatives. Free free to let me know if you have any other ideas 🙂

* Re-implement the styles of the action bar and active box shadow inside the iframe again. We might need to hard-coded something since we cannot get CSS variables inside an iframe, e.g. the border radius of the outer frame, colors, etc. Or we can try to [copy styles](https://bobbyhadz.com/blog/javascript-copy-all-styles-from-one-element-to-another) from one element to another but I'm not familiar with this one.
* Un-iframe if we don't need to fetch the external styles. Currently, if we are passing the inline styles to the iframe, those styles will be wrapped with the `editor-styles-wrapper` class name. Hence, it might be possible to remove the iframe and render the patterns directly.

**Demo**

https://user-images.githubusercontent.com/13596067/234808652-8d72b896-9895-4cc1-8191-e41a64a7e20a.mov

**Sticky action bar**

![image](https://user-images.githubusercontent.com/13596067/234809940-5ab2f100-f2aa-4e13-991b-94b6bcc7d5f2.png)

**Gap and Colors**

![image](https://user-images.githubusercontent.com/13596067/234810080-dc462330-c643-4b57-8f4b-70b5c12a3f50.png)

**Patterns with 100vh height**

| Before | After | Site Editor |
| - | - | - |
| ![image](https://user-images.githubusercontent.com/13596067/234809100-cb956887-9158-4e58-9fc1-e5a9a8df739a.png) | ![image](https://user-images.githubusercontent.com/13596067/234809030-a7e21244-6e4e-4609-89d3-5061d326e130.png) | ![image](https://user-images.githubusercontent.com/13596067/234809200-9e60aa87-bd5d-4dfc-98ef-b8d8302e1d57.png) |
| ![image](https://user-images.githubusercontent.com/13596067/234809384-e502980d-c37c-4b2c-bb36-22349041c795.png) | ![image](https://user-images.githubusercontent.com/13596067/234809448-afbfa968-be70-4080-b33e-971c512e4d5c.png) | ![image](https://user-images.githubusercontent.com/13596067/234809507-b17d4ff8-33c2-4ae4-be26-2e3f4d254955.png) |
| ![image](https://user-images.githubusercontent.com/13596067/234809602-c160e5cf-9543-4f19-bacd-4ee84d7a7b71.png) | ![image](https://user-images.githubusercontent.com/13596067/234809650-06bac578-a180-4b14-8a33-0112b9f94fbc.png) | ![image](https://user-images.githubusercontent.com/13596067/234809696-16e3d7fb-0e02-43e4-b31d-87a99c1915a4.png) |

## Testing Instructions



* Go to /setup/site-setup?siteSlug=<your_free_site>
* Continue until you land on the Design Picker
* On the Design Picker screen, scroll down to the bottom and select BCPA CTA
* On the Pattern Assembler screen
  * Select any header, sections, and footer. Ensure the gap between them looks good
  * Select the pattern with 100vh height, e.g. Link in Bio patterns. Ensure the preview looks good
  * Select any color variation. Ensure the background color looks good. Also, when switching between different color variations, verify the background color of the gap won't be flashing.
  * Hover and scroll the patterns on the large preview. Ensure the action bar and active box shadow works as expected
  * Reorder or remove the patterns via the action bar on the large preview. Ensure the action bar and active box shadow works as expected

## Pre-merge Checklist
